### ISSUE TITLE
Cartesian product + Index cause wrong result

### ISSUE THREAD

**Issue by Aurora-yzj** [memgraph/memgraph#1648]
**Memgraph version**
2.13.0

**Environment**
Operating system: Ubuntu 22.04
Architecture: AMD64
Connect method: any connect method
Driver: any driver
With docker

**Describe the bug**
Memgraph outputs an empty query result when using an index.

**To Reproduce**
1. Run the following query to insert the initial value and create the index:
```
CREATE INDEX ON :L6(k38);
CREATE (n0 :L0:L6{k0 : false, k1 : 890765719, k40 : true, k2 : false, k3 : false, k4 : false, k41 : true, k5 : -1311698154, k6 : "MCotH", k37 : false, k36 : true, k39 : false, id : 0});
CREATE (n0 :L0:L6{k0 : true, k1 : 1304545343, k40 : true, k2 : true, k3 : true, k4 : false, k41 : false, k5 : 1173141482, k6 : "kAIBO99", k37 : true, k36 : true, k39 : false, k38 : 269756692, id : 1});
CREATE (n0 :L0:L6:L11{k0 : false, k70 : "Ml7qmnNQ", k1 : 1098267178, k40 : true, k2 : true, k3 : false, k4 : true, k41 : false, k5 : -786706104, k66 : "QTakNN", k6 : "Wzw0uPB", k65 : 1803875538, k68 : "JeNb8E7V1", k67 : false, k37 : true, k36 : true, k69 : -109922962, k39 : false, k38 : -385421608, id : 2});
CREATE (n0 :L0{k1 : -1006597681, k2 : true, k3 : true, k4 : false, k5 : -1599968686, id : 3, k6 : "xikzo6"});
CREATE (n0 :L0:L6{k1 : 1588792135, k40 : true, k2 : false, k3 : false, k4 : true, k41 : true, k5 : 680249422, k6 : "XF7kSg", k37 : false, k36 : false, k39 : false, k38 : 1463404308, id : 4});
CREATE (n0 :L0:L6{k0 : true, k1 : -1731198578, k40 : true, k2 : true, k3 : false, k4 : true, k41 : false, k5 : -376075495, k6 : "wxjyq", k37 : false, k36 : true, k39 : false, k38 : -1643381344, id : 5});
CREATE (n0 :L0{k0 : true, k1 : -421862127, k2 : false, k3 : true, k4 : false, k5 : -804881356, id : 6, k6 : "5uWHrml5"});
CREATE (n0 :L0{k0 : true, k1 : 1186645042, k2 : true, k3 : false, k4 : false, k5 : 317875177, id : 10, k6 : "RQWjl6f"});
CREATE (n0 :L0{k0 : true, k1 : 1501921859, k2 : false, k3 : false, k4 : true, id : 11, k6 : "Go1qNVk"});
CREATE (n0 {id : 12});
CREATE (n0 :L0:L6{k39 : true, k38 : 584961151, k1 : 483108999, k40 : false, k3 : false, k4 : true, k41 : false, k5 : -1058954276, id : 13, k6 : "MdgjILd6", k37 : false, k36 : true});
CREATE (n0 :L0{k0 : false, k2 : true, k3 : true, k4 : false, k5 : 1026284000, id : 14, k6 : "VV3QWuM"});
MATCH (n0 {id : 11}), (n1 {id : 1}) MERGE(n0)-[r :T2{k82 : true, k84 : "cINRUa", k83 : "dhqqVS", k86 : "wbvvzT", id : 15, k87 : "o2E5gn6Pd"}]->(n1);
MATCH (n0 {id : 1}), (n1 {id : 11}) MERGE(n0)-[r :T2{k82 : false, k84 : "MADM6hN", k83 : "yUnpwg5x", k86 : "WPbrxNw", k85 : "OMqJ8U7Qo", id : 16, k87 : "ESCbuJZC"}]->(n1);
MATCH (n0 {id : 6}), (n1 {id : 13}) MERGE(n0)-[r :T2{k82 : true, k84 : "22vAEWLL", k83 : "sfI4CZ", k86 : "0spkB", k85 : "J43Oj0T", id : 17, k87 : "SqpUzADY6"}]->(n1);
MATCH (n0 {id : 12}), (n1 {id : 2}) MERGE(n0)-[r :T2{k82 : true, k84 : "SXPR8", k83 : "7fEVLMo", k86 : "ec5dz7V", id : 18, k87 : "Nl1Qj"}]->(n1);
MATCH (n0 {id : 4}), (n1 {id : 1}) MERGE(n0)-[r :T2{k82 : true, k84 : "c69H0a", k83 : "mmqOkH7aB", k86 : "qnvkL", k85 : "R6yE23qW", id : 19, k87 : "Ti2tmu"}]->(n1);
MATCH (n0 {id : 4}), (n1 {id : 3}) MERGE(n0)-[r :T2{k82 : true, k84 : "Oc5ADxBQv", k83 : "5MJoJ1", k86 : "goIpN", k85 : "kjHxD", id : 20, k87 : "jJxcsJb"}]->(n1);
MATCH (n0 {id : 5}), (n1 {id : 11}) MERGE(n0)-[r :T2{k82 : false, k84 : "cGmCVNy1I", k83 : "4V8NM6", k86 : "GajNPg", k85 : "bvVqb", id : 21, k87 : "T4O2qE"}]->(n1);
MATCH (n0 {id : 0}), (n1 {id : 4}) MERGE(n0)-[r :T2{k82 : true, k84 : "RBnxH", k83 : "W5bjdNwqI", k86 : "UeNbe", k85 : "ox81Ul3", id : 22, k87 : "mkOiK9O"}]->(n1);
MATCH (n0 {id : 1}), (n1 {id : 5}) MERGE(n0)-[r :T2{k83 : "flRLSv", k86 : "BoyJin", k85 : "1XXjRL", id : 23, k87 : "lRDlTXD"}]->(n1);
MATCH (n0 {id : 10}), (n1 {id : 3}) MERGE(n0)-[r :T2{k82 : false, k84 : "6apbxcZF", k83 : "vZKbZ", k86 : "fwXgFPRw", k85 : "zMMwEQDpL", id : 24, k87 : "MVqgG"}]->(n1);
```
2. Run the query:
```
MATCH (n0 :L6 )<-[r0]-(n1), (n2 :L6)<-[r1]-(n3 :L0) UNWIND [1] AS a0 MATCH (n4)<-[r2]-(n5 :L0 :L6) WHERE (((n2.k1) > (n0.k38)))  RETURN *
```
This query should return some results but it returns an empty result:
![image](https://github.com/memgraph/memgraph/assets/24768784/c49b36e4-8a94-4b74-8022-3399e1f45d72)

If we don't create the index in Step 1 (delete the first line in Step 1), then this query will output the result normally:
![image](https://github.com/memgraph/memgraph/assets/24768784/26eaca68-9fa2-48f4-898f-2a177461f73b)

**Expected behavior**
The query should return normal results with the index created.

Besides, while minimizing the test case, I also found that the UNWIND clause in this query is crucial in reproducing this issue. Hope this will help you locate the root cause. Thank you!

**Verification Environment**
* Plain memgraph Docker image
* Architecture: x86

**Comment by karmenrabar** [comment]
Hello @Aurora-yzj, and thank you for opening this issue! Someone from the team will get back to you.

**Comment by imilinovic** [comment]
@Aurora-yzj If that is a blocker for you, try running memgraph with --cartesian-product-enabled=false, I'm pretty sure the issue won't happen then since this is a problem with cartesian product but your queries memory and speed might be impacted due to not using cartesian.

**Comment by andrejtonev** [comment]
fixed with #1924

**Comment by matea16** [comment]
Hi @Aurora-yzj, I'm happy to let you know that your issue has been closed and resolved with Memgraph's latest version, 2.18 :)
Feel free to test it out and let us know if you notice any futher unexpected behavior. How's your project progressing?
